### PR TITLE
Tests for GroupDivision and SlotAssignment

### PR DESCRIPTION
**Pull request by TLE931**
Tests for GroupDivision and SlotAssignment

GroupBalancedDivision: if-clause to check for non drinkers (ran into exception with no non drinkers)

### CODE REVIEW

File: tests/Feature/SlotAssignmentTest.php [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+        // create a new event
+        $this->event = new Event();
+        $this->event->name = 'slot_booking_test';
+        $this->event->description = 'Lorem ipsum';
+        $this->event->type = 'slot_booking';
+
+        $registration_from = new DateTime(NOW());
+        $registration_to = new DateTime(NOW());
+        $registration_to->modify('+7 days');
```
Comment (simonostendorf) [inline tests/Feature/SlotAssignmentTest.php:35]: Currently all variables are named `camelCase` instead of `snake_case`

File: tests/Feature/SlotAssignmentTest.php [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+            ->count($amount)
+            ->create();
+    }
+
+    /**
+     * A test for the user assignment to slots.
+     */
+    public function testAssignWithMoreRegistrationsThanFreeSlots(): void
+    {
+        $reg_amount = 50;
```
Comment (simonostendorf) [inline tests/Feature/SlotAssignmentTest.php:83]: Currently all variables are named `camelCase` instead of `snake_case`

File: tests/Feature/SlotAssignmentTest.php [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                $this->assertNull($registrations[$i]->queue_position);
+                continue;
+            }
+            $this->assertEquals($i - 9, $registrations[$i]->queue_position);
+        }
+    }
+
+    public function testAssignWithLessRegistrationsThanSlots(): void
+    {
+        $reg_amount = 5;
```
Comment (simonostendorf) [inline tests/Feature/SlotAssignmentTest.php:107]: Currently all variables are named `camelCase` instead of `snake_case`

File: tests/Feature/SlotAssignmentTest.php [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+        $this->assertEquals(0, $this->slot->registrations()->where('queue_position', -1)->get()->count());
+
+        // queue position should be null for all registrations
+        $this->assertEquals($reg_amount, $this->slot->registrations()->where('queue_position', null)->get()->count());
+    }
+
+    public function testAssignWithExactRegistrations(): void
+    {
+        $reg_amount = 10;
```
Comment (simonostendorf) [inline tests/Feature/SlotAssignmentTest.php:121]: Currently all variables are named `camelCase` instead of `snake_case`

File: tests/Feature/SlotAssignmentTest.php [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        $this->assertEquals(0, $this->slot->registrations()->where('queue_position', -1)->get()->count());
+
+        // queue position should be null for all registrations
+        $this->assertEquals($reg_amount, $this->slot->registrations()->where('queue_position', null)->get()->count());
+    }
+
+    public function testReassignAfterNewRegistrations(): void
+    {
+        $reg_amount = 5;
+        $reg_amount_later = 12;
```
Comment (simonostendorf) [inline tests/Feature/SlotAssignmentTest.php:136]: Currently all variables are named `camelCase` instead of `snake_case`

File: database/factories/CourseFactory.php [hunk 6]
```diff
@@ -19,7 +19,7 @@ public function definition(): array
         return [
             'name' => $this->faker->word(),
             'abbreviation' => $this->faker->lexify('????'),
-            'color' => $this->faker->word(),
+            'classes' => $this->faker->word(),
```
Comment (simonostendorf) [inline database/factories/CourseFactory.php:22]: `classes` should not be set to a random word, because they represent tailwind classes.
Instead a valid tailwind-class should be used (see examples in [`CourseSeeder.php`](https://github.com/fsr5-fhaachen/portals/blob/main/database/seeders/CourseSeeder.php)).

File: app/Helpers/GroupBalancedDivision.php [hunk 7]
```diff
@@ -112,35 +112,38 @@ protected function assignNonDrinkers()
...
+                if ($chunks->count() == 0) {
+                    return;
+                }
+                $chunk = $chunks->pop();
+
+                foreach ($chunk as $registration) {
+                    $registration->group_id = $group->id;
+                    $registration->queue_position = null;
+                    $registration->save();
+                }
```
Comment (simonostendorf) [inline app/Helpers/GroupBalancedDivision.php:146]: Could you explain why you changes these lines for implementing a test for this function? :D (this is weird if you think about it)
Is this a bugfix in the algorithm?

File: README.md [hunk 8]
```diff
@@ -97,10 +97,16 @@ npm run lint:fix
 
 ### testing
 
+Copy phpunit.xml.example to phpunit.xml and fill in the test database credentials or leave it as is for in-memory testing with sqlite.
```
Comment (simonostendorf) [inline README.md:100]: @TobiasGrether do you think it would make sense to update the dependencies and libs to support testing with [Laravel Pest](https://pestphp.com/)? (I think this is the modern solution instead of phpunit)

**Review by simonostendorf — CHANGES_REQUESTED** [review]
I added some comments and change requests. 

Currently no time to test the tests locally.

File: app/Helpers/GroupBalancedDivision.php (hunk 7 above)
Comment (simonostendorf) [reply app/Helpers/GroupBalancedDivision.php:146]: Sorry, now i saw your comment.

> GroupBalancedDivision: if-clause to check for non drinkers (ran into exception with no non drinkers)

Comment (simonostendorf) [reply app/Helpers/GroupBalancedDivision.php:146]: Does this maybe also fix #503 ?

**Comment by simonostendorf** [comment]
The tests must run in the ci run (see https://github.com/fsr5-fhaachen/portals/blob/dev/.github/workflows/ci.yml#L65-L69)

File: tests/Feature/SlotAssignmentTest.php [hunk 9]
```diff
@@ -104,58 +104,45 @@ public function testAssignWithMoreRegistrationsThanFreeSlots(): void
...
-        $registrations = $this->slot->registrations()->orderBy('queue_position')->get();
-        for ($i = 0; $i < $registrations->count(); $i++) {
-            if ($i < 10) {
-                $this->assertNull($registrations[$i]->queue_position);
-                continue;
-            }
-            $this->assertEquals($i - 9, $registrations[$i]->queue_position);
-        }
-
-        $this->createUsers(100);
```
Comment (simonostendorf) [inline tests/Feature/SlotAssignmentTest.php]: Why did you remove these lines in the "variable naming" commit? :)

File: README.md (hunk 8 above)
Comment (TobiasGrether) [reply README.md:100]: @simonostendorf PHPUnit is generally decent, I've only used it for static code analysis so far. But yeah Laravel Pest would probably be a better choice. API looks nice

File: app/Helpers/GroupBalancedDivision.php (hunk 7 above)
Comment (TobiasGrether) [reply app/Helpers/GroupBalancedDivision.php:146]: > Does this maybe also fix #503 ?

Maybe yeah. Nice catch

**Comment by TobiasGrether** [comment]
@jpih @TLE931 Do the tests all run to successful completion for you? I tried to run them locally and two of them still seem to error out

`There were 2 failures:

1) Tests\Feature\SlotAssignmentTest::testAssignWithMoreRegistrationsThanFreeSlots
Failed asserting that 1 is null.

/var/www/html/tests/Feature/SlotAssignmentTest.php:98
/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Testing/TestCase.php:61

2) Tests\Feature\SlotAssignmentTest::testReassignAfterNewRegistrations
Failed asserting that 1 is null.

/var/www/html/tests/Feature/SlotAssignmentTest.php:152
/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Testing/TestCase.php:61

FAILURES!
Tests: 10, Assertions: 4498, Failures: 2.
`

File: app/Helpers/GroupBalancedDivision.php (hunk 7 above)
Comment (TLE931) [reply app/Helpers/GroupBalancedDivision.php:146]: I don't think so. I think the imbalance in this issue is because assigning non-drinkers takes priority over dividing users by courses equally. This could lead to said imbalance in some cases

File: README.md (hunk 8 above)
Comment (TLE931) [reply README.md:100]: I tried migrating to pest, but the tests won't run due to problems with the faker in factories. I have no idea how to fix this.

**Comment by TLE931** [comment]
> The tests must run in the ci run (see https://github.com/fsr5-fhaachen/portals/blob/dev/.github/workflows/ci.yml#L65-L69)

how do I tell the CI where the phpunit.xml is?

**Comment by TobiasGrether** [comment]
> > The tests must run in the ci run (see https://github.com/fsr5-fhaachen/portals/blob/dev/.github/workflows/ci.yml#L65-L69)
> 
> how do I tell the CI where the phpunit.xml is?
> 
> edit: phpunit.xml is in gitignore. I think it is safe to not ignore, because the tests run on in-memory db, so the configuration works for everyone.

You can also rename the phpunit.xml.example to phpunit.xml, I suppose since its no longer going to be just examples after this gets merged. You can remove the phpunit.xml from the .gitignore, it doesn't contain any credentials and is needed for the automated test runs

**Comment by TLE931** [comment]
das war aus Versehen, hat sich noch nichts geändert

**Comment by TLE931** [comment]
Closing this because branch got mangled. PR for new branch: https://github.com/fsr5-fhaachen/portals/pull/677
